src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/TemplatesTests.cs

```diff
@@ -0,0 +1,137 @@
+                Assert.Equal("artifacts", ((CollectGCDumpOptions)options.Actions[0].Settings).Egress);
+                Assert.Equal(KnownCollectionRuleActions.CollectGCDump, options.Actions[1].Type);
+                Assert.Equal("artifacts2", ((CollectGCDumpOptions)options.Actions[1].Settings).Egress);
+                Assert.Equal(KnownCollectionRuleActions.CollectTrace, options.Actions[2].Type);
+                Assert.Equal("monitorBlob", ((CollectTraceOptions)options.Actions[2].Settings).Egress);
+                Assert.Equal(WebApi.Models.TraceProfile.Cpu, ((CollectTraceOptions)options.Actions[2].Settings).Profile);
+                Assert.Equal(KnownCollectionRuleActions.CollectDump, options.Actions[3].Type);
+                Assert.Equal("monitorBlob", ((CollectDumpOptions)options.Actions[3].Settings).Egress);
+
+                // Filters Comparison
+                Assert.Equal(2, options.Filters.Count);
+                Assert.Equal(WebApi.ProcessFilterKey.ProcessName, options.Filters[0].Key);
+                Assert.Equal("FirstWebApp", options.Filters[0].Value);
+                Assert.Equal(WebApi.ProcessFilterKey.ProcessName, options.Filters[1].Key);
+                Assert.Equal("FirstWebApp1", options.Filters[1].Value);
+                Assert.Equal(WebApi.ProcessFilterType.Exact, options.Filters[1].MatchType);
+
+                // Limits Comparison
+                Assert.Equal(1, options.Limits.ActionCount);
+                Assert.Equal(TimeSpan.Parse("00:00:30"), options.Limits.ActionCountSlidingWindowDuration);
+                Assert.Equal(TimeSpan.Parse("00:05:00"), options.Limits.RuleDuration);
+
+            }, builder: (HostBuilder)builder);
+        }
+
+        /// <summary>
+        /// Tests that incorrectly referenced Templates cause options to be nullifed.
+        /// </summary>
+        [Fact]
+        public async void TemplatesTranslationFailTest()
+        {
+            using TemporaryDirectory userConfigDir = new(_outputHelper);
+
+            IHostBuilder builder = GetHostBuilder(userConfigDir);
+
+            await TestHostHelper.CreateCollectionRulesHost(_outputHelper, rootOptions => { }, host =>
+            {
+                IOptionsMonitor<CollectionRuleOptions> optionsMonitor = host.Services.GetRequiredService<IOptionsMonitor<CollectionRuleOptions>>();
+
+                OptionsValidationException ex = Assert.Throws<OptionsValidationException>(() => optionsMonitor.Get("InvalidRule"));
+
+                string[] failures = ex.Failures.ToArray();
+                Assert.Equal(2, failures.Length);
+
+                Assert.Equal(string.Format(TemplateNotFoundErrorMessage, "TriggerTemplateINVALID"), failures[0]);
+                Assert.Equal(string.Format(TemplateNotFoundErrorMessage, "FilterTemplateINVALID"), failures[1]);
+
+            }, builder: (HostBuilder)builder);
+        }
+
+        private IHostBuilder GetHostBuilder(TemporaryDirectory userConfigDir)
```

Comment:
I guess I'm not sure what the process is for converting the current approach to using `overrideSource`, but what's the downside of the current approach (other than adding another parameter)? I like that it currently follows the pattern we use in `ConfigurationTests`, but it doesn't need to if there's a reason not to.

Reply:
> I guess I'm not sure what the process is for converting the current approach to using `overrideSource`, but what's the downside of the current approach (other than adding another parameter)?

You are trying to add a new parameter that has an already existing solution that shouldn't be painful to use. If there is no actual benefit for the alternative, then the alternatives shouldn't be added.

For how to use `overrideSource`, you can try (1) deserialize the JSON doc, traverse it and flatten it into a dictionary (key is configuration key path, value is value), or (2) refactor `overrideSource` to be an `IConfigurationSource` that is just added rather than calling AddInMemoryCollection, change the existing usage to pass `MemoryConfigurationSource`, and have your new usage pass a `JsonConfigurationSource` or `JsonStreamConfigurationSource` (the latter should let you use a MemoryStream over your JSON data rather than having to write it disk).